### DISCUSSION TITLE
Can regex_syntax support the parser that only parses ASCII character patterns?

### DISCUSSION THREAD

**Discussion by ybbh** [rust-lang/regex#1177] [Q&A] [answered]
I want to build a regex parser that only parses ASCII character patterns.
For example, I want to parse the ASCII string without the newline "\n" character in the following source code.
Is there a function like ".ascii(true)" to allow this action?
 
```
fn main() {
  let pattern = "[^\n]*";
  
  let mut parser = regex_syntax::ParserBuilder::new().unicode(false).build();
  
  // This line code complains of an Error of invalid utf8 pattern.
  // Is there a function like .ascii(true) to allow the following code line to pass?
  let _r = parser.parse(pattern).unwrap();
}
```

**Comment by BurntSushi** [comment, marked as answer]
There is no `ascii` mode. The error you're getting occurs because, when Unicode mode is disabled, `[^\n]` matches _any byte_ except for `\n`. This includes bytes like `\xFF`, which are neither ASCII nor valid UTF-8.

An "ASCII" mode is really just a subset of Unicode mode. So it's best to leave Unicode mode enabled, but change your pattern to match ASCII exclusively. For example, to match the set of ASCII codepoints that aren't `\n`, you could write `[\p{ascii}&&[^\n]]`.

**Reply by ybbh** [reply]
Thank you for your quick reply. I see, so ASCII match with Unicode mode is the best way.
